File: js/StaticAsteroids.js
function StaticAsteroids(num, ctx) {
    var x;
    var y;
    var nextPt = function(pt) {
        var sign = Math.random() < 0.5 ? -1 : 1;
        return pt + (sign * Math.random() * 15);
    };
    ctx.clearRect(0,0,canvas.width,canvas.height);
    for(i = 0; i < num; i++){
        x = Math.random() * canvas.width;
        y = Math.random() * canvas.height;
        ctx.beginPath();
        ctx.moveTo(x,y);
        for(j=0; j < 10; j++){
            x = nextPt(x);
            y = nextPt(y);
            ctx.lineTo(x, y);
        }
        ctx.closePath();
        ctx.stroke();
    }
}

File: hs/StaticAsteroids.hs
{-# LANGUAGE ApplicativeDo    #-}
{-# LANGUAGE ParallelListComp #-}
module StaticAsteroids (benchmark, summary) where

import           Control.Monad  hiding (sequence_)
import           Data.Foldable  (sequenceA_)
import           Graphics.Blank
import           System.Random
import           Utils

benchmark :: CanvasBenchmark
benchmark ctx = do
  xs  <- replicateM numAsteroids $ randomXCoord ctx
  ys  <- replicateM numAsteroids $ randomYCoord ctx
  dxs <- replicateM numAsteroids $ randomRIO (-15, 15)
  dys <- replicateM numAsteroids $ randomRIO (-15, 15)
  send' ctx $ do
             clearCanvas
             sequenceA_ [showAsteroid (x,y) (mkPts (x,y) ds)
                       | x <- xs
                       | y <- ys
                       | ds <- cycle $ splitEvery 6 $ zip dxs dys
                       ]

summary :: String
summary = "StaticAsteroids"

numAsteroids :: Int
numAsteroids = 1000

showAsteroid :: Point -> [Point] -> Canvas ()
showAsteroid (x,y) pts = do
  beginPath()
  moveTo (x,y)
  mapM_ lineTo pts
  closePath()
  stroke()

mkPts :: Point -> [(Double, Double)] -> [Point]
mkPts (x,y) xs = [ (x+x',y+y') | (x',y') <- xs ]

splitEvery :: Int -> [a] -> [[a]]
splitEvery n = takeWhile (not.null) . map (take n) . iterate (drop n) -- borrowed from Magnus Kronqvist on stack overflow
